src/Compilers/CSharp/Portable/BoundTree/Statement.cs

```diff
@@ -310,20 +320,49 @@ partial class BoundStateMachineScope
         protected override OperationKind StatementKind => OperationKind.None;
     }
 
-    partial class BoundLocalDeclaration : IVariableDeclarationStatement, IVariable
+    partial class BoundLocalDeclaration : IVariableDeclarationStatement
     {
-        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables => ImmutableArray.Create<IVariable>(this);
-
-        ILocalSymbol IVariable.Variable => this.LocalSymbol;
+        private static readonly ConditionalWeakTable<BoundLocalDeclaration, IVariable> s_variablesMappings =
+            new ConditionalWeakTable<BoundLocalDeclaration, IVariable>();
 
-        IExpression IVariable.InitialValue => this.InitializerOpt;
+        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables
+        {
+            get
+            {
+                var builder = ArrayBuilder<IVariable>.GetInstance(1);
+                var variable = s_variablesMappings.GetValue(this, decl => new VariableDeclaration(decl.LocalSymbol, decl.InitializerOpt, decl.Syntax));
+                builder.Add(variable);
+                return builder.ToImmutableAndFree();
```

Comment:
Can we cache an array in order to avoid allocating it over and over again? Also, from the API perspective, it is probably expected that property returns the same array every time.